### PR TITLE
[clusters] Network Commissioning Server

### LINKED ISSUE #3945: Add network provisioning cluster to CHIP
#### Problem
per spec https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/309, we need initial network provisioning cluster xml

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
We don't have network commissioning cluster on the server side.

 #### Summary of Changes
Add clusters for network commissioning to the lighting app server side. After this is merged and tested, will scale to all example apps.

Also, the AddThreadNetwork temporary does not work, it will be fixed by full support of bytes string.

 #### Test
```
chip-device-ctrl > connect -ip fd12::2 12345678
...
Device temporary node id (**this does not match spec**): 269170
chip-device-ctrl > zcl NetworkCommissioning AddWiFiNetwork 269170 1 0 ssid=test credentials=12345678 breadcrumb=0 timeoutMs=1000
...
chip-device-ctrl > zcl NetworkCommissioning EnableNetwork 269170 1 0 networkID=test breadcrumb=0 timeoutMs=1
...
```

The device can connect to the network now.

 Fixes #3945

### CODE REVIEW

File: examples/lighting-app/lighting-common/gen/CHIPClustersObjc.h
Comment (andy31415) [inline examples/lighting-app/lighting-common/gen/CHIPClustersObjc.h:97]: is this typical style? 

Maybe at least add a newline in between method declarations. I find this odd to read (but I do not read a lot of objc)

Would some documentation be useful? Like what is a 'breadcrumb'

File: examples/lighting-app/lighting-common/gen/CHIPClustersObjc.mm
Comment (andy31415) [inline examples/lighting-app/lighting-common/gen/CHIPClustersObjc.mm:627]: would a `NO_MEMORY` error be more correct that incorrect state?
Same below.

File: examples/lighting-app/lighting-common/gen/CHIPClustersObjc.mm
Comment (andy31415) [inline examples/lighting-app/lighting-common/gen/CHIPClustersObjc.mm:640]: in case of failure, are onSuccess/onFailure auto-cleared?

I am trying to understand why we only free memory on failure since it sseems like there is no delete on success.
Same for below.

File: src/app/clusters/network-commissioning/network-commissioning-ember.cpp
Comment (andy31415) [inline src/app/clusters/network-commissioning/network-commissioning-ember.cpp:34]: These doccomments seem useless to me.

Could we remove them from the template or do they have a chance to contain something useful?

File: src/app/clusters/network-commissioning/network-commissioning-ember.cpp
Comment (andy31415) [inline src/app/clusters/network-commissioning/network-commissioning-ember.cpp:101]: what do the true/false callback mean? 

It looks as if when having code they return true, otherwise false. Some constants wouold be nice like 'handled' or 'nothandled'.

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (andy31415) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:31]: why this odd #define followed by inclusion of the define?

can't we just include directly? if this is because macro expansion, please add a comment explaining.

Could this be resolved in some different way by using the build system?

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (andy31415) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:64]: do we need kLabel here? What do they mean?

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (andy31415) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:120]: this will permanently use up some RAM - is this the intented final usage or are we supposed to migrate to some central 'connectivity' object to manage these?

We currently do not need this to join wifi at boot, so I imagine there should be a way too not need this.

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (andy31415) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:127]: why is the datatype not const char * to start with? Are we guaranteed that this is a null terminated string (and if so, where?)

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (andy31415) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:156]: nit: can declare these where needed, to reduce scoping (more specific locality is easier to read)

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (andy31415) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:178]: nit: I believe the exit pattern in this case makes the code harder to read. This could be:

```cpp
for (size_t i = 0; i < kMaxNetworks; i++) 
{
    if (sNetworks[i].mNetworkType != NetworkType::kUndefined) {
        continue;
    }
    
    memcpy(sNetworks[i].mData.mWiFi.mSSID, ssid, ssid_Len);
￼    sNetworks[i].mData.mWiFi.mSSIDLen = ssid_Len;

    // ....
    
    ChipLogDetail(Zcl, "WiFi provisioning data: SSID: %s", ssid);
    return CHIP_NO_ERROR;
}
 
return CHIP_ERROR_NO_MEMORY;
```

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning.cpp:120]: we supposed to migrate to some central 'connectivity' object to manage these. In fact, I'd like to move these to attribute storage if possible in the future updates.

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning.cpp:127]: According to the spec, this should be a OCTET_STRING, it should be a uint8_t*. Currently, the octet_string is not handled by current code gen correctly, I may add this support later (as described in the TODOs at the beginning of this file.

File: examples/lighting-app/lighting-common/gen/CHIPClustersObjc.h
Comment (erjiaqing) [reply examples/lighting-app/lighting-common/gen/CHIPClustersObjc.h:97]: @vivien-apple Could you add a document about how we are using ObjC with CHIP library?

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning.cpp:129]: Updated, thanks.

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning.cpp]: This is defined in the command definition, I realized that these enums should be generated by code gen so I removed them.

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning.cpp:33]: I would say we should remove DeviceNetworkProvisioningDelegateImpl put provide some platform neutral api in connectivity manager or some place else.
Currently we have other places (the rendezvous session) that is using DeviceNetworkProvisioningDelegateImpl, so I basically use this but I'd like to deprecate DeviceNetworkProvisioningDelegateImpl after we fully switched to network provisioning cluster.

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning.cpp]: Thanks.

File: src/app/clusters/network-commissioning/network-commissioning-ember.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning-ember.cpp]: They are just implementing callbacks required by the ember library. I think removing this is totally ok, however, I also want to make the network commissioning a template of how to write cluster handlers, so I left some comments here and removed the notes for params -- they are just params for cluster commands.

File: src/app/clusters/network-commissioning/network-commissioning-ember.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning-ember.cpp:96]: Updated comments for this.

File: examples/lighting-app/lighting-common/gen/CHIPClustersObjc.h
Comment (vivien-apple) [reply examples/lighting-app/lighting-common/gen/CHIPClustersObjc.h:97]: I do not have a document handy, but all the view in https://github.com/project-chip/connectedhomeip/tree/master/src/darwin/CHIPTool/CHIPTool/View%20Controllers uses the Objc APIs.

About the documentation itself, we need to fix ZAP to retrieve some documentation from the `xml` definition. At the moment there is no field in the SQL database for it and I believe the JS parser needs to be updated too. @jepenven-silabs  do you have any plans or work in the pipe to add support for that ?

File: examples/lighting-app/lighting-common/gen/CHIPClustersObjc.mm
Comment (vivien-apple) [reply examples/lighting-app/lighting-common/gen/CHIPClustersObjc.mm:640]: > in case of failure, are onSuccess/onFailure auto-cleared?
> 
> I am trying to understand why we only free memory on failure since it sseems like there is no delete on success.
> Same for below.

"in case of failure" == "in case of success" ? If so, the callback the callback is cleared once it has been fired (e.g  https://github.com/project-chip/connectedhomeip/blob/master/src/darwin/Framework/CHIP/gen/CHIPClustersObjc.mm#L50)

Now that I'm looking at it again I think one of the callback is leaking though since only one of them can fire. I will definitively fix it in an update to the API. Thanks !

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (vivien-apple) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:105]: That's not really the error I guess. This is `BoundsExceeded` right ?

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (vivien-apple) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:170]: Maybe `CHIP_ERROR err = CHIP_ERROR_KEY_NOT_FOUND` instead of setting the value later

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (vivien-apple) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:178]: Maybe use a `NetworkInfo *` to temporary hold `sNetworks[networkSeq]` that would be easier to read

Comment (vivien-apple) [reply src/app/clusters/network-commissioning/network-commissioning.cpp:178]: You also needs to update the `mEnabled` state of the `NetworkInfo` struct per spec.

File: src/app/clusters/network-commissioning/network-commissioning-ember.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning-ember.cpp:26]: Fixed, Thanks

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning.cpp:203]: Alghouth it should be updated via attribute storage, I just update it in DoEnableNetwork as a reminder.

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (erjiaqing) [reply src/app/clusters/network-commissioning/network-commissioning.cpp]: Made a change for changing all functions in this file "Internal", and then made them return the enum type for errors.

**Review by andy31415 — APPROVED** [review]

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (vivien-apple) [inline src/app/clusters/network-commissioning/network-commissioning.cpp:130]: Maybe `err = OutOfRange`.

**Review by vivien-apple — APPROVED** [review]
Approving assuming the error code will be changed before merging.

**Review by jelderton — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/app/clusters/network-commissioning/network-commissioning.cpp
Comment (andy31415) [reply src/app/clusters/network-commissioning/network-commissioning.cpp]: octet_string should have a length then - null_terminated_octed_string sounds odd.

**Comment by yuzibo** [comment]
Hi, pull latest code and compiling  chip-device-ctrl:
```bash
zcl  NetworkCommissioning AddWiFiNetwork 269170 1 0 ssid=test credentials=12345678 breadcrumb=0 timeoutMs=1000
An exception occurred during process ZCL command:
Parsing Error: value should be encoded in encoding:encodedvalue format
```
No clue to solve it.

**Comment by erjiaqing** [comment]
> Hi, pull latest code and compiling chip-device-ctrl:
> 
> ```shell
> zcl  NetworkCommissioning AddWiFiNetwork 269170 1 0 ssid=test credentials=12345678 breadcrumb=0 timeoutMs=1000
> An exception occurred during process ZCL command:
> Parsing Error: value should be encoded in encoding:encodedvalue format
> ```
> 
> Test it in many ways. No clue to solve it.

zcl  NetworkCommissioning AddWiFiNetwork 269170 1 0 ssid=str:test credentials=str:12345678 breadcrumb=0 timeoutMs=1000

I will add some more documents about this.

**Pull request merged**
